File: java/stone_game_1.java
class Solution {
    public boolean stoneGame(int[] piles) {
        int n = piles.length;
        int [][]dp = new int[n+1][n+1];
 
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j] = -1;
            }
        }
        
        return solve(piles, dp, 0, n-1) > 0;
    }
    
    public int solve(int []piles, int [][]dp, int i, int j) {
        if (i > j) {
            return 0;
        }
        
        if (dp[i][j] != -1) {
            return dp[i][j];
        }
        
        dp[i][j] = Math.max(piles[i] - solve(piles, dp, i+1, j), piles[j] - solve(piles, dp, i, j-1));
        return dp[i][j];
    }
}

File: js/stone_game_1.js
/**
 * @param {number[]} piles
 * @return {boolean}
 */
 var stoneGame = function(piles) {
    let n = piles.length;
    let dp = [...Array(n)].map(_ => Array(n).fill(-1));
    
    let solve = (i = 0, j = n - 1) => {
        if (i > j) {
            return 0;
        }
        
        if (dp[i][j] != -1) {
            return dp[i][j];
        }
      
        dp[i][j] = Math.max(piles[i] - solve(i + 1, j), piles[j] - solve(i, j - 1));
        return dp[i][j];
    };
    return solve() > 0;
};